File: dotnet/src/Microsoft.Agents.Protocol.Validation/Validators/ContentValidatorBase.cs
using Microsoft.Agents;

namespace Microsoft.Agents.Protocol.Validation.Validators;

/// <summary>
/// Base class for content validators.
/// </summary>
/// <typeparam name="T">The type of content to validate.</typeparam>
public abstract class ContentValidatorBase<T> : IValidator<T> where T : AIContent
{
    /// <summary>
    /// Validates the content and returns a validation result.
    /// </summary>
    public abstract ValidationResult Validate(T content, ValidationContext? context = null);

    /// <summary>
    /// Helper method to create a validation error.
    /// </summary>
    protected ValidationError CreateError(string code, string message, string? field = null)
    {
        return new ValidationError(code, message, field);
    }

    /// <summary>
    /// Helper method to validate that a string is not empty.
    /// </summary>
    protected ValidationError? ValidateNotEmpty(string? value, string fieldName, string errorCode, string errorMessage)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return CreateError(errorCode, errorMessage, fieldName);
        }
        return null;
    }

    /// <summary>
    /// Helper method to validate that a number is positive.
    /// </summary>
    protected ValidationError? ValidatePositive(int? value, string fieldName, string errorCode, string errorMessage)
    {
        if (value.HasValue && value.Value <= 0)
        {
            return CreateError(errorCode, errorMessage, fieldName);
        }
        return null;
    }

    /// <summary>
    /// Helper method to validate that a number is positive.
    /// </summary>
    protected ValidationError? ValidatePositive(long? value, string fieldName, string errorCode, string errorMessage)
    {
        if (value.HasValue && value.Value <= 0)
        {
            return CreateError(errorCode, errorMessage, fieldName);
        }
        return null;
    }

    /// <summary>
    /// Helper method to validate that a number is positive.
    /// </summary>
    protected ValidationError? ValidatePositive(float? value, string fieldName, string errorCode, string errorMessage)
    {
        if (value.HasValue && value.Value <= 0)
        {
            return CreateError(errorCode, errorMessage, fieldName);
        }
        return null;
    }

    /// <summary>
    /// Helper method to validate JSON format.
    /// </summary>
    protected ValidationError? ValidateJson(string? json, string fieldName, string errorCode, string errorMessage)
    {
        if (string.IsNullOrWhiteSpace(json))
            return null;

        try
        {
            System.Text.Json.JsonDocument.Parse(json);
            return null;
        }
        catch (System.Text.Json.JsonException)
        {
            return CreateError(errorCode, errorMessage, fieldName);
        }
    }

    /// <summary>
    /// Helper method to validate URI format.
    /// </summary>
    protected ValidationError? ValidateUri(string? uri, string fieldName, string errorCode, string errorMessage)
    {
        if (string.IsNullOrWhiteSpace(uri))
            return null;

        if (!Uri.TryCreate(uri, UriKind.Absolute, out _))
        {
            return CreateError(errorCode, errorMessage, fieldName);
        }
        return null;
    }
}

File: typescript/packages/agents-protocol-validation/src/validators/base/ContentValidatorBase.ts
import type { AIContent } from '@microsoft/agents-protocol-abstractions';
import type { IValidator } from '../../core/IValidator';
import type { ValidationContext } from '../../core/ValidationContext';
import { ValidationResult } from '../../core/ValidationResult';
import { ValidationError } from '../../core/ValidationError';

/**
 * Base class for all content validators.
 * Provides common validation helper methods.
 */
export abstract class ContentValidatorBase<T extends AIContent> implements IValidator<T> {
  /**
   * Validates the content object.
   */
  public abstract validate(obj: T, context?: ValidationContext): ValidationResult;

  /**
   * Validates that a string value is non-empty.
   * @returns ValidationError if validation fails, undefined otherwise
   */
  protected validateNotEmpty(
    value: string | null | undefined,
    fieldName: string,
    errorCode: string,
    errorMessage: string
  ): ValidationError | undefined {
    if (!value || value.trim().length === 0) {
      return new ValidationError(errorCode, errorMessage, fieldName);
    }
    return undefined;
  }

  /**
   * Validates that a numeric value is positive (> 0).
   * @returns ValidationError if validation fails, undefined otherwise
   */
  protected validatePositive(
    value: number | null | undefined,
    fieldName: string,
    errorCode: string,
    errorMessage: string
  ): ValidationError | undefined {
    if (value !== null && value !== undefined && value <= 0) {
      return new ValidationError(errorCode, errorMessage, fieldName);
    }
    return undefined;
  }

  /**
   * Validates that a numeric value is non-negative (>= 0).
   * @returns ValidationError if validation fails, undefined otherwise
   */
  protected validateNonNegative(
    value: number | null | undefined,
    fieldName: string,
    errorCode: string,
    errorMessage: string
  ): ValidationError | undefined {
    if (value !== null && value !== undefined && value < 0) {
      return new ValidationError(errorCode, errorMessage, fieldName);
    }
    return undefined;
  }

  /**
   * Validates that a string is valid JSON.
   * @returns ValidationError if validation fails, undefined otherwise
   */
  protected validateJson(
    json: string | null | undefined,
    fieldName: string,
    errorCode: string,
    errorMessage: string
  ): ValidationError | undefined {
    if (!json) {
      return undefined; // Empty is OK unless checked separately
    }

    try {
      JSON.parse(json);
      return undefined;
    } catch {
      return new ValidationError(errorCode, errorMessage, fieldName);
    }
  }

  /**
   * Validates that a string is a valid URI.
   * @returns ValidationError if validation fails, undefined otherwise
   */
  protected validateUri(
    uri: string | null | undefined,
    fieldName: string,
    errorCode: string,
    errorMessage: string
  ): ValidationError | undefined {
    if (!uri) {
      return undefined; // Empty is OK unless checked separately
    }

    try {
      new URL(uri);
      return undefined;
    } catch {
      return new ValidationError(errorCode, errorMessage, fieldName);
    }
  }

  /**
   * Validates that a string matches a regex pattern.
   * @returns ValidationError if validation fails, undefined otherwise
   */
  protected validatePattern(
    value: string | null | undefined,
    pattern: RegExp,
    fieldName: string,
    errorCode: string,
    errorMessage: string
  ): ValidationError | undefined {
    if (!value) {
      return undefined; // Empty is OK unless checked separately
    }

    if (!pattern.test(value)) {
      return new ValidationError(errorCode, errorMessage, fieldName);
    }

    return undefined;
  }

  /**
   * Validates that a string does not exceed a maximum length.
   * @returns ValidationError if validation fails, undefined otherwise
   */
  protected validateMaxLength(
    value: string | null | undefined,
    maxLength: number,
    fieldName: string,
    errorCode: string,
    errorMessage: string
  ): ValidationError | undefined {
    if (value && value.length > maxLength) {
      return new ValidationError(errorCode, errorMessage, fieldName);
    }
    return undefined;
  }

  /**
   * Validates that a value is within a set of allowed values.
   * @returns ValidationError if validation fails, undefined otherwise
   */
  protected validateAllowedValues<V>(
    value: V | null | undefined,
    allowedValues: V[],
    fieldName: string,
    errorCode: string,
    errorMessage: string
  ): ValidationError | undefined {
    if (value !== null && value !== undefined && !allowedValues.includes(value)) {
      return new ValidationError(errorCode, errorMessage, fieldName);
    }
    return undefined;
  }

  /**
   * Helper to collect non-null validation errors.
   */
  protected collectErrors(...errors: (ValidationError | undefined)[]): ValidationError[] {
    return errors.filter((e): e is ValidationError => e !== undefined);
  }
}
